### ISSUE TITLE
Inconsistent paths when installing dependency with a different name

### ISSUE THREAD

**Issue by apaleslimghost** [bower/bower#2135]
When explicitly installing a package that is registered under multiple names, the package is always placed in `bower_components` under the canonical name from the package's bower.json. For example, `bower install quartertobowertest-a` and `bower install quartertobowertest-b` both install to `bower_components/quartertobowertest-a`.

However, if another package has it in `dependencies`, it is installed to the name listed in dependencies no matter what. For example, `bower install quartertobowertest-dep`, which has:

``` json
"dependencies": {
  "quartertobowertest-b": "*"
}
```

installs the package to `bower_components/quartertobowertest-b`. This can cause duplicate installs when multiple packages depend on the package under different names, or the root package depends on the package and its dependent.

This was initially reported as ftlabs/ftdomdelegate#77, which is registered as `dom-delegate` in the main registry and `ftdomdelegate` in the [Origami registry](http://registry.origami.ft.com/components).

**Comment by sheerun** [comment]
It's interesting. Maybe you can write a failing test in `test/commands/install.js` ?

**Comment by BenMann** [comment]
I'm not so sure this is something we can fix soon. I'll summarize the situation: 

Because the bower.json is set to `name: dom-delegate` and there is only one (the same) repo for both packages the following happens: 

Bower registry > `dom-delegate` > resolves to: `dom-delegate`
Origami registry > `ftdomdelegate` > resolves to: `dom-delegate` 
Which is by design, as the packages are literally pointing to the same repo only named differently and the bower.json is the single point of truth here. Not surprising — agree? 
> Quick solution to fix this confusion: rename in Bower registry & bower.json ==> both packages now resolve to the same repo and have the same name. All good, you solved all problems. Then again you now have the same packages twice.. get rid of one!
   
When installing as a dependency you specify the name, which means Bower does (so it seems) not check the registry, but assumes you entered the right name and fetches exactly that package, installing it as specified in the depency-section (it does not resolve the name from the bower.json). This is the real problem and it only occurs because you have two packages which in reality are one. Even pointing to the same repo. The easy solution: again.. don't have two packages that really are one! That way you will not have duplicate installs from different depencies.    
 
The problem behind the scenes: If more than one registry is in use (or actually always), we need to resolve dependency-names from the bower.json.   
 Using just one registry, we also need to make sure we do not allow to add a package that points to a repo which is already registered under a different name! I suspect right now this can also happen, as `bower lookup jquery` and `bower lookup jQuery` might point to the same repo. As dependencies then, they are different packages (not tested).   
One can prevent this by making the DB fields for `name` and `url` unique in our registry, however it might be too late — will need to find examples here.
   
Just out of curiosity: what's the reason to have the same package in both registries?

**Comment by rajzshkr** [comment]
@BenMann I saw plugins like jquery also registered many times, This may happen if the user trying to check bower register is happening or not?  So we have to make both **name** and **URL** unique, because we should not allow same package with different name and different package with same also. If you confirm I will crack the code of registry this week end.

**Comment by BenMann** [comment]
TLDR: If you have a better solution than described in :two:, I'm all for it!

We should not touch the registry for the actual issue described, as it wouldn't fix that problem. There can already be cases like this and it is now a little delicate which ever way we go, because we don't want to break current builds. Let's divide this into two before going into detail..

-
Here's an easy testcase (that actually works):
We have three packages. Two point to the same repo: `bowernametesta` and `bowernametestb`. They are effectivley the same package (same repo), registered under two names. --> :one:
A third package which is called `bowernametestreq` requires this dependency:
```
"dependencies": {
    "bowernametestb": "bowernametestb#*"   
  }
```
Now if you install those two packages:
`bower install bowernametesta` and then
 `bower install bowernametestreq`
you installed the same package under different names because the dependency is listed under a different name in `bowernametestreq`. --> :two:

-

I'll start with :two: : Dependencies are always installed to `/bower_components/<name in "dependencies">`, even though the `bower.json` of that package shows a different name. That's the actual problem. A possible solution is to check the `bower.json` of every dependency that will be installed and install it to `/bower_components/<name in dependencies' bower.json>`. This however might break current builds if we change it (cc: @sheerun riiight)? Without having looked at the code, such a fix is probably somewhere between `install` and resolvers.

:one: is a different beast that I would totally postpone to a new registry version. What really needs to be checked is, whether a `url` is already present in the database and if this is the case, throw an error and inform the user that the package is already registered under name xyz. Easy for one DB but ideally this would magically work between different registries :smile:.

**Comment by BenMann** [comment]
I'm going to reference [this issue](https://github.com/bower/bower/issues/2067) here, because :one: is actually even worse, if those packages are named identically, but one is lower-case and the other camel-case.

**Comment by sheerun** [comment]
Maybe in Bower 2.0 we can ignore registry aliases and use name from bower.json

**Comment by sheerun** [comment]
Feature like this won't be implemented because bower is in maintenance mode now
